rest/replicator_test.go

```diff
@@ -3070,3 +3072,275 @@ func TestBlipSyncNonUpgradableConnection(t *testing.T) {
 	require.NoError(t, err, "Error sending request")
 	require.Equal(t, http.StatusUpgradeRequired, response.StatusCode)
 }
+
+// TestActiveReplicatorPullConflictReadWriteIntlProps:
+//   - Starts 2 RestTesters, one active, and one passive.
+//   - Create the same document id with different content on rt1 and rt2
+//   - Publishes the REST API on a httptest server for the passive node (so the active can connect to it)
+//   - Uses an ActiveReplicator configured for pull to start pulling changes from rt2.
+func TestActiveReplicatorPullConflictReadWriteIntlProps(t *testing.T) {
+
+	bodyBytes := func(body string) []byte {
+		buffer := new(bytes.Buffer)
+		require.NoError(t, json.Compact(buffer, []byte(body)))
+		return buffer.Bytes()
+	}
+
+	// scenarios
+	conflictResolutionTests := []struct {
+		name               string
+		localRevisionBody  db.Body
+		localRevID         string
+		remoteRevisionBody db.Body
+		remoteRevID        string
+		conflictResolver   string
+		expectedLocalBody  db.Body
+		expectedLocalRevID string
+	}{
+		{
+			name:               "mergeReadWriteIntlProps",
+			localRevisionBody:  db.Body{"source": "local"},
+			localRevID:         "1-a",
+			remoteRevisionBody: db.Body{"source": "remote"},
+			remoteRevID:        "1-b",
+			conflictResolver: `function(conflict) {
+				var mergedDoc = new Object();
+				mergedDoc.source = "merged";
+				mergedDoc.incomingDocId = conflict.RemoteDocument._id;
+				mergedDoc.incomingRevId = conflict.RemoteDocument._rev;
+				mergedDoc.outgoingDocId = conflict.LocalDocument._id;
+				mergedDoc.outgoingRevId = conflict.LocalDocument._rev;
+				return mergedDoc;
+			}`,
+			expectedLocalBody: db.Body{
+				"incomingDocId": "mergeReadWriteIntlProps",
+				"incomingRevId": "1-b",
+				"outgoingDocId": "mergeReadWriteIntlProps",
+				"outgoingRevId": "1-a",
+				"source":        "merged",
+			},
+			expectedLocalRevID: db.CreateRevIDWithBytes(2, "1-b", bodyBytes(`{
+				"incomingDocId":"mergeReadWriteIntlProps",
+				"incomingRevId":"1-b",
+				"outgoingDocId":"mergeReadWriteIntlProps",
+				"outgoingRevId":"1-a",
+				"source":"merged"}`)),
+		},
+		{
+			name: "mergeReadWriteAttachments",
+			localRevisionBody: map[string]interface{}{
+				"_attachments": map[string]interface{}{"A": map[string]interface{}{"data": "QQo="}},
+				"source":       "local",
+			},
+			localRevID: "1-a",
+			remoteRevisionBody: map[string]interface{}{
+				"_attachments": map[string]interface{}{"B": map[string]interface{}{"data": "Qgo="}},
+				"source":       "remote",
+			},
+			remoteRevID: "1-b",
+			conflictResolver: `function(conflict) {
+				var mergedDoc = new Object();
+				mergedDoc.source = "merged";
+				dst = conflict.RemoteDocument._attachments;
+				src = conflict.LocalDocument._attachments;
+				for (var key in src) { dst[key] = src[key]; }
+				mergedDoc._attachments = dst;
```

Comment:
Would be best to avoid modifying the value in dst to produce a union of both attachments. Instead consider:

```suggestion
				var mergedAtts = new Object();
				for (var key in dst) {
					mergedAtts[key] = dst[key];
				}
				for (var key in src) {
					mergedAtts[key] = src[key];
				}
				mergedDoc._attachments = mergedAtts;
```

Reply:
This has been committed in latest revision